src/net-mdns/src/ServiceDiscovery.cs

```diff
@@ -0,0 +1,468 @@
+            {
+                log.Debug($"Answer from {e.RemoteEndPoint}");
+            }
+            if (log.IsTraceEnabled)
+            {
+                log.Trace(msg);
+            }
+
+            // Any DNS-SD answers?
+
+            var sd = msg.Answers
+                .OfType<PTRRecord>()
+                .Where(ptr => ptr.Name.IsSubdomainOf(LocalDomain));
+            foreach (var ptr in sd)
+            {
+                if (ptr.Name == ServiceName)
+                {
+                    ServiceDiscovered?.Invoke(this, ptr.DomainName);
+                }
+                else if (ptr.TTL == TimeSpan.Zero)
+                {
+                    var args = new ServiceInstanceShutdownEventArgs
+                    {
+                        ServiceInstanceName = ptr.DomainName,
+                        Message = msg
+                    };
+                    ServiceInstanceShutdown?.Invoke(this, args);
+                }
+                else
+                {
+                    var args = new ServiceInstanceDiscoveryEventArgs
+                    {
+                        ServiceInstanceName = ptr.DomainName,
+                        Message = msg
+                    };
+                    ServiceInstanceDiscovered?.Invoke(this, args);
+                }
+            }
+        }
+
+        void OnQuery(object sender, MessageEventArgs e)
+        {
+            var request = e.Message;
+
+            if (log.IsDebugEnabled)
+            {
+                log.Debug($"Query from {e.RemoteEndPoint}");
+            }
+            if (log.IsTraceEnabled)
+            {
+                log.Trace(request);
+            }
+
+            // Determine if this query is requesting a unicast response
+            // and normalise the Class.
+            var QU = false; // unicast query response?
+            foreach (var r in request.Questions)
+            {
+                if (((ushort)r.Class & 0x8000) != 0)
+                {
+                    QU = true;
+                    r.Class = (DnsClass)((ushort)r.Class & 0x7fff);
+                }
+            }
+
+            var response = NameServer.ResolveAsync(request).Result;
+
+            if (response.Status != MessageStatus.NoError)
+            {
+                return;
+            }
+
+            // Many bonjour browsers don't like DNS-SD response
+            // with additional records.
+            if (response.Answers.Any(a => a.Name == ServiceName))
+            {
+                response.AdditionalRecords.Clear();
+            }
+
+            if (AnswersContainsAdditionalRecords)
+            {
+                response.Answers.AddRange(response.AdditionalRecords);
+                response.AdditionalRecords.Clear();
+            }
+
+            if (!response.Answers.Any(a => a.Name == ServiceName))
+            {
+                ;
+            }
+
+            if (QU)
+            {
+                // TODO: Send a Unicast response if required.
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)